src/software/new_geom/util/intersection.h

```diff
@@ -0,0 +1,76 @@
+#include "software/geom/util.h"
+#include "software/new_geom/ray.h"
+#include "software/new_geom/rectangle.h"
+#include "software/new_geom/segment.h"
+
+/**
+ * Computes the intersection of two segments.
+ *
+ * @param first
+ * @param second
+ *
+ * @return  one of:
+ *          - an empty vector if no intersections
+ *          - a vector containing a single point of intersection
+ *          - a vector containing two points representing the line segment of the overlap
+ * if both segments are collinear and overlapping
+ */
+std::vector<Point> intersection(const Segment &first, const Segment &second);
+
+/**
+ * Computes the points of intersection between a rectangle and a line segment.
+ *
+ * @param rectangle
+ * @param segment
+ *
+ * @return a set containing the points of intersection between a rectangle and a
+ * segment
+ */
+std::set<Point> intersection(const Rectangle &rectangle, const Segment &segment);
+
+/**
+ * Computes the points of intersection between a ray and a line segment.
+ *
+ * @param ray
+ * @param segment
+ *
+ * @return  one of:
+ *          - an empty vector if no intersections
+ *          - a vector containing a single point of intersection
+ *          - a vector containing two points representing the line segment of the overlap
+ * if the ray and segment are collinear and overlapping
+ */
+std::vector<Point> intersection(const Ray &ray, const Segment &segment);
+
+/**
+ * Computes the point of intersection between two lines.
+ *
+ * @pre the lines are not parallel
```

Comment:
Same as above

Reply:
Same thing, I removed it and added a note that overlapping lines arent considered as intersecting